### PR TITLE
Fix compiling ObjC/ObjC++ on Windows/MinGW

### PR DESCRIPTION
**Pull request by Midar**

### CODE REVIEW

**Comment by amyspark** [comment]
Hi @Midar, I just finished tracking down why the `24 mixing languages` test consistently fails on MSVC. It's a mix of issues:

1. On the Meson side, there's a mix of CL (for C) and Clang (for Objective-C). Neither of them implements the `__STDC__` macro for STL compatibility reasons ([source](https://learn.microsoft.com/en-us/cpp/build/reference/zc-stdc?view=msvc-170), [source](https://stackoverflow.com/a/66966733)). This macro is a requirement for CMake's [language standard detection support under MSVC](https://github.com/Kitware/CMake/blob/aa6b2180bf9926e62641871908ae059c19ddb3a3/Modules/CMakeOBJCCompilerId.m.in#L32-L47).
2. The CMake test does not set languages up correctly, causing `OBJC` files to be built as `C` (and thus `cl`).

To fix (1), it's enough to add below here:

https://github.com/mesonbuild/meson/blob/43b80e02ce0e87dfcf069111e62ad8eff4435d6e/mesonbuild/cmake/toolchain.py#L187-L190

the following block:

```python
            if lang.startswith('objc') and comp_obj.get_id().startswith('clang'):
                defaults[f'{prefix}FLAGS'] = ['-D__STDC__=1']
```

That will uncover (2), and to fix that https://github.com/mesonbuild/meson/blob/master/test%20cases/cmake/24%20mixing%20languages/subprojects/cmTest/CMakeLists.txt#L3 needs to read

```cmake
project(cmTest LANGUAGES C OBJC)
```

This will uncover an issue in CMake: it cannot maintain two sets of argument syntaxes.

```
| CMake Error at C:/Program Files/CMake/share/cmake-3.27/Modules/CMakeTestOBJCCompiler.cmake:64 (message):
| The Objective-C compiler

| "C:/Program Files/Microsoft Visual Studio/2022/Preview/VC/Tools/Llvm/x64/bin/clang.EXE"

| is not able to compile a simple test program.

| It fails with the following output:

| Change Dir: 'E:/meson/test cases/cmake/24 mixing languages/build/subprojects/cmTest/__CMake_build/CMakeFiles/CMakeScratch/TryCompile-hr7wdl'

| Run Build Command(s): C:/ProgramData/chocolatey/bin/ninja.exe -v cmTC_f1714
| [1/2] C:\PROGRA~1\MIB055~1\2022\Preview\VC\Tools\Llvm\x64\bin\clang.EXE   -x objective-c -D__STDC__=1 -MD -MT CMakeFiles/cmTC_f1714.dir/testObjCCompiler.m.o -MF CMakeFiles\cmTC_f1714.dir\testObjCCompiler.m.o.d -o CMakeFiles/cmTC_f1714.dir/testObjCCompiler.m.o -c "E:/meson/test cases/cmake/24 mixing languages/build/subprojects/cmTest/__CMake_build/CMakeFiles/CMakeScratch/TryCompile-hr7wdl/testObjCCompiler.m"
| [2/2] cmd.exe /C "cd . && C:\PROGRA~1\MIB055~1\2022\Preview\VC\Tools\Llvm\x64\bin\clang.EXE -D__STDC__=1 /machine:x64 CMakeFiles/cmTC_f1714.dir/testObjCCompiler.m.o -o cmTC_f1714.exe   && cd ."
| FAILED: cmTC_f1714.exe
| cmd.exe /C "cd . && C:\PROGRA~1\MIB055~1\2022\Preview\VC\Tools\Llvm\x64\bin\clang.EXE -D__STDC__=1 /machine:x64 CMakeFiles/cmTC_f1714.dir/testObjCCompiler.m.o -o cmTC_f1714.exe   && cd ."
| clang: error: no such file or directory: '/machine:x64'
| ninja: build stopped: subcommand failed.
```

File: test cases/cmake/24 mixing languages/meson.build [hunk 1]
```diff
@@ -7,6 +7,11 @@ if not add_languages('objc', required : false)
   error('MESON_SKIP_TEST: No ObjC compiler')
 endif
 
+objc = meson.get_compiler('objc')
+if objc.get_define('_MSC_VER') != ''
+  error('MESON_SKIP_TEST: fixme: cmake must get the correct compiler when mixing languages')
+endif
+
```
Comment (amyspark) [inline test cases/cmake/24 mixing languages/meson.build:14]: If you want, once the fixes in my comment are applied, you can document why it actually doesn't work:

```suggestion
objc = meson.get_compiler('objc')
c = meson.get_compiler('c')
if c.get_argument_syntax() != objc.get_argument_syntax()
  error('MESON_SKIP_TEST: cmake cannot mix compiler types on windows')
endif
```

**Comment by dcbaker** [comment]
@amyspark, @Midar: this all looks good to me, so if we can get Amyspark's changes in I'd be happy to get this merged.

**Comment by Midar** [comment]
@amyspark Thank you so much, your suggestions have been incorporated!

CI is running now.

**Comment by Midar** [comment]
@dcbaker CI passed, so this is ready to merge.

How would we make this end up in a 1.5.2 release?

**Review by dcbaker — APPROVED** [review]

**Comment by eli-schwartz** [comment]
> The CMake test does not set languages up correctly, causing `OBJC` files to be built as `C` (and thus `cl`).

We just hit this in https://bugs.gentoo.org/938133 as well, which is on Linux. GCC is built without objc support, clang has it, so all CFLAGS are applied to cmTest.m.

This then fails if your CFLAGS happen to contain GCC-specific arguments that clang doesn't accept...

**Comment by eli-schwartz** [comment]
> How would we make this end up in a 1.5.2 release?

I think I agree, yes. Tagging for inclusion in the next bugfix (whenever that should be).

**Pull request merged**